org-code-javabuilder/validation/src/main/java/org/code/validation/support/NeighborhoodTracker.java

```diff
@@ -0,0 +1,98 @@
+package org.code.validation.support;
+
+import static org.code.protocol.ClientMessageDetailKeys.*;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.code.neighborhood.NeighborhoodSignalKey;
+import org.code.neighborhood.World;
+import org.code.protocol.ClientMessage;
+import org.code.protocol.ClientMessageType;
+import org.code.validation.NeighborhoodLog;
+import org.code.validation.PainterEvent;
+import org.code.validation.PainterLog;
+import org.code.validation.Position;
+
+public class NeighborhoodTracker {
+  private final Map<String, PainterTracker> painterTrackers;
+  private String[][] neighborhoodState;
+  private boolean isInitialized;
+
+  public NeighborhoodTracker() {
+    this.painterTrackers = new HashMap<>();
+    this.isInitialized = false;
+  }
+
+  public NeighborhoodLog getNeighborhoodLog() {
+    final ArrayList<PainterTracker> trackers = new ArrayList<>(painterTrackers.values());
+    final PainterLog[] painterLogs = new PainterLog[trackers.size()];
+    for (int i = 0; i < painterLogs.length; i++) {
+      painterLogs[i] = trackers.get(i).getPainterLog();
+    }
+    return new NeighborhoodLog(painterLogs, this.neighborhoodState);
+  }
+
+  public void trackEvent(ClientMessage message) {
+    if (message.getType() != ClientMessageType.NEIGHBORHOOD) {
+      return;
+    }
+
+    final String id = message.getDetail().getString(ID);
+    final NeighborhoodSignalKey key = NeighborhoodSignalKey.valueOf(message.getValue());
+    if (key == NeighborhoodSignalKey.INITIALIZE_PAINTER) {
+      if (!this.isInitialized) {
+        this.initializeGrid();
+      }
+      final int x = message.getDetail().getInt(X);
+      final int y = message.getDetail().getInt(Y);
+      final int paint = message.getDetail().getInt(PAINT);
+      final PainterTracker painterTracker = new PainterTracker(id, new Position(x, y), paint);
+      this.painterTrackers.put(id, painterTracker);
+      return;
+    }
+
+    if (!this.painterTrackers.containsKey(id) || !this.isInitialized) {
+      System.out.printf(
+          "Received a painter event for an uninitialized painter. Ignoring message: %s, %s\n",
+          message.getValue(), message.getDetail());
```

Comment:
Wanted to ask about this - should we throw an exception instead?

Reply:
this would indicate something wrong with our initialization code--a student shouldn't be able to trigger painter events without first initializing a painter and calling events on it. I would maybe have this be an error log so we get notified something weird is happening.

Reply:
Cool makes sense. Do you think we should throw as well so the validation run stops, or should we continue with validation?

Reply:
Either way would be fine. If we had this bug on our end the student also would see weird behavior when they tried to run their code (because we would be trying to move/paint with an nonexistent painter) so I think we can just assume we won't see an event for a painter that doesn't exist. I would lean towards continue with validation in case things can pass, so that students don't see a bug on our end that doesn't impact them.